### PR TITLE
Implement display list-based plot resizing for notebooks

### PR DESCRIPTION
**Pull request by jmcphers**
This change improves the appearance of plots in notebooks in three ways:

1. Plots are now drawn for the notebook's exact width. Plots were formerly drawn at a fixed width and height and scaled to the editor surface using traditional image scaling.

2. The device pixel ratio is now respected, so plots appear sharper on Retina displays. 

3. When the editor is resized, plots are redrawn to fit the new width from a persisted display list.

Display lists are persisted as `.snapshot` files inside the notebook cache; the display lists are written along with the plot's image. A resize causes the image to be recomputed from the display list.

To avoid tying up the main thread during resize, the resizing is performed by a separate R process, which receives the list of `.snapshot` files on standard input and outputs the resized `.png` filenames to standard output, where they're picked up and sent to the client as they become available.

### CODE REVIEW

File: src/cpp/session/modules/rmarkdown/NotebookPlots.cpp [hunk 1]
```diff
@@ -105,9 +109,42 @@ void processPlots(const FilePath& plotFolder,
...
+   // if there's a plot on the device, write its display list before it's
+   // cleared for the next page
+   FilePath outputFile = pPlotState->plotFolder.complete(
+         core::system::generateUuid(false) + kDisplayListExt);
+   Error error = r::exec::RFunction(".rs.saveGraphics", 
+         outputFile.absolutePath()).call();
+
+   // if there's already an unconsumed display list, remove it, since this
+   // display list replaces it
+   if (pPlotState->snapshotFile.empty())
```
Comment (jjallaire) [inline src/cpp/session/modules/rmarkdown/NotebookPlots.cpp]: Should this be `!pPlotState->snapshotFile.empty()` ?

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/TextEditingTargetNotebook.java [hunk 2]
```diff
@@ -837,10 +878,51 @@ public void onResize(ResizeEvent event)
...
       }
       
-      for (ChunkOutputUi output: outputs_.values())
+      // lightly debounce local resizes since they're somewhat expensive
+      resizePlotsLocal_.schedule(50);
+      
+      // heavily debounce remote resizes since they're very expensive
+      // (this actually spins up a separate R process to re-render all the
+      // plots at the new resolution)
+      resizePlotsRemote_.schedule(500);
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/TextEditingTargetNotebook.java]: Does this mean we wait until 500ms after the user finishes sizing to resize the plots? (that seems desirable).

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/TextEditingTargetNotebook.java [hunk 3]
```diff
@@ -1490,6 +1572,75 @@ public void run()
...
+      public void run()
+      {
+         for (ChunkOutputUi output: outputs_.values())
+         {
+            output.getOutputWidget().syncHeight(false, false);
+         }
+      }
+   };
+   
+   private Timer resizePlotsRemote_ = new Timer()
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/TextEditingTargetNotebook.java]: To further underline my previous comment, I'm thinking that if it takes e.g. 3 seconds for the user to resize that there should be one resize event that happens at the end.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/TextEditingTargetNotebook.java (hunk 2 above)
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/rmd/TextEditingTargetNotebook.java]: Yes, that's correct; it's 500ms measured from the time the user finishes resizing (as long as the size is actively changing the timer won't fire).

File: src/cpp/session/modules/rmarkdown/NotebookPlots.cpp (hunk 1 above)
Comment (jmcphers) [inline src/cpp/session/modules/rmarkdown/NotebookPlots.cpp]: Nice catch. Fixed!

**Pull request merged**